jax/_src/scipy/special.py

```diff
@@ -657,3 +661,290 @@ def i1e(x):
+    The 3D array representing the derivatives of associated Legendre functions
+    of the first kind.
+  """
+
+  num_m, num_l, num_x = p.shape
+
+  # p_{l-1}^m.
+  p_m_lm1 = jnp.pad(p, ((0, 0), (1, 0), (0, 0)))[:, :num_l, :]
+
+  # p_{l-1}^{m+2}.
+  p_mp2_lm1 = jnp.pad(p_m_lm1, ((0, 2), (0, 0), (0, 0)))[2:num_m + 2, :, :]
+
+  # p_{l-1}^{m-2}.
+  p_mm2_lm1 = jnp.pad(p_m_lm1, ((2, 0), (0, 0), (0, 0)))[:num_m, :, :]
+
+  # Derivative computation requires negative orders.
+  if is_normalized:
+    raise NotImplementedError(
+        'Negative orders for normalization is not implemented yet.')
+  else:
+    if num_l > 1:
+      l_vec = jnp.arange(1, num_l - 1)
+      p_p1 = p[1, 1:num_l - 1, :]
+      coeff = -1.0 / ((l_vec + 1) * l_vec)
+      update_p_p1 = jnp.einsum('i,ij->ij', coeff, p_p1)
+      p_mm2_lm1 = p_mm2_lm1.at[ops.index[1, 2:num_l, :]].set(update_p_p1)
+
+    if num_l > 2:
+      l_vec = jnp.arange(2, num_l - 1)
+      p_p2 = p[2, 2:num_l - 1, :]
+      coeff = 1.0 / ((l_vec + 2) * (l_vec + 1) * l_vec)
+      update_p_p2 = jnp.einsum('i,ij->ij', coeff, p_p2)
+      p_mm2_lm1 = p_mm2_lm1.at[ops.index[0, 3:num_l, :]].set(update_p_p2)
+
+  m_mat, l_mat = jnp.mgrid[:num_m, :num_l]
+
+  coeff_zeros = jnp.zeros((num_m, num_l))
+  upper_0_indices = jnp.triu_indices(num_m, 0, num_l)
+  zero_vec = jnp.zeros((num_l,))
+
+  with np.errstate(divide='ignore', invalid='ignore'):
+    a0 = -0.5 / (m_mat - 1.0)
+  a0_masked = coeff_zeros.at[upper_0_indices].set(a0[upper_0_indices])
+  a0_masked = a0_masked.at[1, :].set(zero_vec)
+
+  b0 = l_mat + m_mat
+  with np.errstate(invalid='ignore'):
+    c0 = a0 * (b0 - 2.0) * (b0 - 1.0)
+  c0_masked = coeff_zeros.at[upper_0_indices].set(c0[upper_0_indices])
+  c0_masked = c0_masked.at[1, :].set(zero_vec)
+
+  # p_l^{m-1}.
+  p_mm1_l = (jnp.einsum('ij,ijk->ijk', a0_masked, p_m_lm1) +
+             jnp.einsum('ij,ijk->ijk', c0_masked, p_mm2_lm1))
+
+  d0 = -0.5 / (m_mat + 1.0)
+  d0_masked = coeff_zeros.at[upper_0_indices].set(d0[upper_0_indices])
+  e0 = d0 * b0 * (b0 + 1.0)
+  e0_masked = coeff_zeros.at[upper_0_indices].set(e0[upper_0_indices])
+
+  # p_l^{m+1}.
+  p_mp1_l = (jnp.einsum('ij,ijk->ijk', d0_masked, p_mp2_lm1) +
+             jnp.einsum('ij,ijk->ijk', e0_masked, p_m_lm1))
+
+  f0 = b0 * (l_mat - m_mat + 1.0) / 2.0
+  f0_masked = coeff_zeros.at[upper_0_indices].set(f0[upper_0_indices])
+  p_derivative = jnp.einsum('ij,ijk->ijk', f0_masked, p_mm1_l) - 0.5 * p_mp1_l
+
+  # Special treatment of the singularity at m = 1.
+  if num_m > 1:
+    l_vec = jnp.arange(num_l)
+    g0 = jnp.einsum('i,ij->ij', (l_vec + 1) * l_vec, p[0, :, :])
+    if num_l > 2:
+      g0 = g0 -  p[2, :, :]
+    p_derivative_m0 = jnp.einsum('j,ij->ij', 0.5 / jnp.sqrt(1 - x * x), g0)
+    p_derivative = p_derivative.at[1, :, :].set(p_derivative_m0)
+    p_derivative = p_derivative.at[1, 0, :].set(jnp.zeros((num_x,)))
+
+  return p_derivative
+
+
+@partial(jit, static_argnums=(0, 2))
+def gen_associated_legendre(l_max: int,
```

Comment:
preceding underscore for private function

Reply:
Done. Thanks!